Implement IOList BIFs

mlwilkerson opened the pull request:
Implementation for #153.
- [x] Implement and test `iolist_to_binary/1`
- [x] Implement and test `iolist_size/1`
- [x] Implement and test `iolist_to_iovec/1`
- [x] Add testing for `SubBinary`
- [x] Figure out whether to add testing for `MatchContext`
    This doesn't seem to be possible yet.

mlwilkerson commented:
I'm starting with pretty naive implementations here, to get my feet wet. This is my first Rust programming, and first exposure to both the Lumen dev environment and to iolist, so I've gotta find a handhold somewhere.

It's taken a bit of digging to find good information on the use cases for iolist vs the similar `list_to_binary/1`. (My initial implementations delegate to `list_to_binary/1`.)

I'm beginning to understand the use case of optimizing memory usage when streaming data such as when writing to a file or socket. The iolist functions can allow for more efficient list operations and memory usage. Handling improper lists is part of this use case. Helpful resources on this: [here](https://www.bignerdranch.com/blog/elixir-and-io-lists-part-1-building-output-efficiently/), [here](https://www.bignerdranch.com/blog/elixir-and-io-lists-part-2-io-lists-in-phoenix/), [here](https://elixirforum.com/t/weird-sasl-lists-and-can-i-rely-on-kernel-is-list/470/2), [here](https://stackoverflow.com/questions/36946064/elixir-what-kind-of-list-is-this-with-the-vertical-bar-still-in-it/36946418#36946418).

mlwilkerson commented:
I'm not sure whether adding test coverage for the binary sub-types is important. Interested to hear what a reviewer thinks about that.

KronicDeth commented on lumen_runtime/src/otp/erlang/iolist_to_binary_1/test.rs:
```diff
@@ -0,0 +1,56 @@
+use crate::otp;
+use crate::scheduler::with_process;
+
+// > Bin1 = <<1,2,3>>.
```
❤️  including Erlang doc examples as tests

KronicDeth commented on lumen_runtime/src/otp/erlang/iolist_to_iovec_1.rs:
```diff
@@ -0,0 +1,40 @@
+                    if term.is_binary() {
+                        term
+                    } else {
+                        otp::erlang::list_to_binary_1::native(process, term).unwrap()
```
You can't `unwrap` here.  If a term that is a badarg for `list_to_binary/1` is passed, this will cause a `panic!`.

mlwilkerson replied:
This should be fixed in my latest changes.

mlwilkerson commented on lumen_runtime/src/otp/erlang/iolist_size_1.rs:
```diff
@@ -0,0 +1,21 @@
+/// Returns the size, in bytes, of the binary that would be result from iolist_to_binary/1
+#[native_implemented_function(iolist_to_binary/1)]
+pub fn native(process: &Process, iolist_or_binary: Term) -> exception::Result {
+    let bin = otp::erlang::iolist_to_binary_1::native(process, iolist_or_binary)?;
```
Thanks, that's the insight I needed on this use case. It definitely seemed like a naive initial implementation.

So now I'm thinking that an implementation that reduces the iolist to a byte size would be preferable (i.e. not allocating memory for all it, just scanning through the data structure and accumulating the size required along the way). I'll move in that direction.

mlwilkerson replied:
I've re-written this one and just pushed it.

I think this puts it on the right track.

mlwilkerson commented on lumen_runtime/src/otp/erlang/iolist_size_1/test.rs:
```diff
@@ -0,0 +1,73 @@
+}
+
+#[test]
+fn with_binary_returns_binary() {
```
whoops, probably

mlwilkerson commented:
@KronicDeth do you have any feedback regarding whether it's important to test with various binaries that would result in different internal typed terms? (ProcBin, HeapBinary, SubBinary, MatchContext)

Seems like I saw something possibly related on #200 .

KronicDeth commented:
You definitely need to test subbinaries to show that you handle non-binary ones correctly.

mlwilkerson commented:
I rebased onto `develop` and fixed up some stuff that seems to have changed.

I was not successful at constructing a MatchContext that I could pass in as a term from the test module.

I tried adding a test for ProcBin, which seemed to expose the fact that the `list_to_binary/1` (to which `iolist_to_binary/1` currently delegates). Opened #368 to track that and proposed #369 to resolve it. So the resolution of this PR is now blocking and that resolution of those.
